### PR TITLE
Panel-Method Inviscid Cp as Physics-Informed Input Feature

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**This is our most radical experiment yet — a paradigm shift from everything else we've tried.**

After 17 consecutive failures modifying architecture, loss, optimization, and augmentation, the core insight is: the model doesn't have the right INFORMATION. The NACA6416 fore-foil is OOD, and its tandem wake interaction creates pressure patterns the model has never interpolated. But a simple inviscid panel method can solve the exact same tandem geometry analytically — the Cp distribution is a deterministic function of geometry + AoA, independent of the training data distribution.

**Proposed:** Pre-compute inviscid pressure coefficient (Cp) using a vortex panel solver for every sample. Feed this as an additional input feature channel. The model then predicts the viscous correction (Δp = p_viscous - p_inviscid) rather than pressure from scratch — a much smaller, smoother signal.

**Key evidence:**
- **B-GNN (arXiv:2503.18638)**: Adding panel-method Cp + local Re_x as input features achieved **88% OOD error reduction** on single-foil cases
- **NeuralFoil (arXiv:2503.16323)**: Panel method baseline + neural viscous correction architecture
- The tandem case benefits even MORE since the panel method natively captures the fore-aft aerodynamic interaction (downwash, circulation) — exactly what the network struggles with

## Instructions

This is a multi-step experiment. Take your time to get the panel solver right — correctness matters more than speed.

### Phase 1: Implement Panel Solver (do this FIRST, before any training)

Implement a simple 2-element Hess-Smith vortex panel method in Python. The solver needs to:
1. Take two airfoil geometries (fore + aft), their positions (gap, stagger), and AoA as input
2. Output the inviscid Cp at each surface panel (node)

**Simplest approach — use an existing library if available:**
```python
# Check if any panel method library is installed:
# pip install aeropy  OR  pip install pyvlm  OR implement from scratch
```

**If implementing from scratch, here's the core algorithm:**
```python
def panel_method_tandem(foil1_coords, foil2_coords, gap, stagger, aoa_rad):
    """
    Hess-Smith panel method for 2-element configuration.
    foil1_coords, foil2_coords: [N, 2] numpy arrays of (x, y) for each foil
    Returns: cp1 [N1], cp2 [N2] — pressure coefficients at each panel
    """
    # 1. Position foil2 at (gap, stagger) relative to foil1 trailing edge
    # 2. Build panel geometry (normals, tangents, control points)
    # 3. Compute influence coefficients (source + vortex) for all panels
    # 4. Solve linear system for source strengths + vortex strength
    # 5. Compute Cp = 1 - (V_tangential / V_freestream)^2
    # Key: both foils share the same vortex sheet — the tandem interaction
    # is captured through the mutual influence coefficients
    ...
```

A simpler alternative: use **thin-airfoil theory** with a Kutta condition for each foil. Less accurate but much simpler and still captures the tandem interaction signal.

**Validation:** Run the panel solver on a few known cases (e.g., NACA0012 at 5° AoA) and verify Cp matches textbook values.

### Phase 2: Pre-compute and Cache Panel Cp

```python
# For each sample in training and validation data:
# 1. Extract foil1 geometry, foil2 geometry, gap, stagger, AoA
# 2. Run panel solver
# 3. Interpolate panel Cp to mesh surface nodes (use nearest-neighbor or linear interp)
# 4. Save to .npy cache keyed by sample index

# Cache structure: panel_cp_cache[sample_idx] = np.array of shape [N_nodes]
# For volume nodes (not on surface): set panel_cp = 0.0
```

Store cache in the working directory. Pre-computation for all samples should take <30 minutes.

### Phase 3: Add Panel Cp as Input Feature

1. **Add CLI flag:** `--panel_cp_feature` (action='store_true')
2. **Load panel Cp cache** at dataset init
3. **Append panel_cp as feature channel 25** (or wherever next available) to the input tensor. The model currently has 24 input features — this adds a 25th.
4. **Update the model's input projection** from 24 → 25 dimensions (find where `in_channels` or `n_input` is defined and add 1 when `--panel_cp_feature` is set)

### Phase 4: Train and Evaluate

Run **2 seeds (42, 73)** with the panel Cp feature:

```bash
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/panel-cp-s42" --wandb_group thorfinn-panel-cp \
  --panel_cp_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

**Important notes:**
- If the panel solver is too complex for tandem foils, fall back to **single-foil panel Cp for each foil separately** (ignoring the interaction). Even this provides a physics-informed baseline that the model can learn to correct.
- The panel Cp should be computed in the SAME coordinate frame as the CFD data.
- If pre-computation takes too long, compute panel Cp only for validation samples first and test with a single seed to verify the feature helps before caching all training data.
- Log the correlation between panel Cp and actual Cp at validation time to verify the feature is informative.

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42), j9btfx09 (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Panel-Method Inviscid Cp as Physics-Informed Input Feature

**Implementation:** Hess-Smith vortex panel method solving each foil independently (without tandem interaction). Panel Cp pre-computed for all 1322 training samples and cached to `.npz`. Panel Cp appended as input channel 25 (x dim: 24 → 25). Model `fun_dim` updated accordingly.

### Metrics (Surface MAE pressure)

| Metric | Baseline (2-seed avg) | Panel Cp S42 | Panel Cp S73 | Panel Cp Avg | Delta |
|--------|----------------------|-------------|-------------|-------------|-------|
| p_in | 13.05 | 13.0 | 13.6 | 13.30 | +0.25 (+1.9%) |
| **p_tan** | **28.60** | **30.1** | **29.2** | **29.65** | **+1.05 (+3.7%)** |
| p_oodc | 7.70 | 7.5 | 7.7 | 7.60 | -0.10 (-1.3%) |
| p_re | 6.55 | 6.5 | 6.4 | 6.45 | -0.10 (-1.5%) |

### Detailed Metrics (S42 / S73)

| Metric | S42 | S73 |
|--------|-----|-----|
| val_loss (3-split) | 0.3881 | 0.3848 |
| mae_surf_Ux (in) | 3.03 | 3.25 |
| mae_surf_Uy (in) | 0.85 | 0.86 |
| mae_surf_p (in) | 13.0 | 13.6 |
| mae_vol_p (in) | 12.3 | 12.9 |
| Best epoch | 149 | 150 |
| Peak VRAM | 46.6 GB | 45.4 GB |
| Epoch time | ~72s | ~71s |
| Total time | 180.9 min | 181.1 min |

### W&B Runs
- S42: `xdxcew1g` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xdxcew1g
- S73: `7j0gg5jb` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/7j0gg5jb

### Commands Used

```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent thorfinn \
  --wandb_name "thorfinn/panel-cp-s42" --wandb_group thorfinn-panel-cp \
  --panel_cp_feature --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```
(Same for S73 with `--seed 73`, `CUDA_VISIBLE_DEVICES=1`)

### What happened

**Negative result.** The panel Cp input feature did not improve any critical metric. The key target metric (tandem transfer, p_tan) worsened by 3.7% from 28.60 → 29.65.

**Why it likely failed:**

1. **No tandem interaction modeled.** The panel solver computed each foil independently — the crucial fore-aft aerodynamic interaction (downwash, wake influence) that makes tandem cases hard was NOT captured. A proper 2-element panel method would provide the interaction signal, but that's significantly more complex.

2. **Inviscid Cp is already implicitly learnable.** The model already receives full geometry coordinates, NACA parameters, AoA, and signed distance fields. With its attention-based architecture, it can plausibly learn an implicit Cp approximation from these features alone. Adding an explicit but approximate Cp doesn't provide new information.

3. **Feature normalization mismatch.** Panel Cp values range from ~-3 to +1 and were normalized with mean=0, std=1 (since panel Cp is not part of the original dataset stats). This may have created a feature that the network partially ignored or that interfered with established feature correlations.

4. **Slight VRAM increase** (46.6 vs ~45 GB baseline) suggesting the extra channel adds some overhead but nothing dramatic.

**Note:** Both runs show `nan` in the verification denormalization check. This appears to be a pre-existing issue with the verification code for val_ood_re, not caused by the panel Cp feature.

### Suggested follow-ups

1. **Proper 2-element panel method** — If the hypothesis is to be revisited, a coupled 2-element Hess-Smith solver (with both foils' influence matrices interacting) could capture the tandem interaction signal. The B-GNN paper's 88% OOD error reduction was on single-foil cases; tandem cases need the interaction modeled.

2. **Panel Cp as a residual target** rather than input feature — Instead of adding Cp as input channel 25, train the model to predict `p_viscous - p_inviscid` (the viscous correction). This changes the prediction target rather than the input features and may be easier for the network to learn.

3. **Focus on data augmentation** — The tandem transfer gap may be better addressed by data augmentation strategies that create synthetic tandem variations, rather than physics-informed features that are approximate.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2179 because ADVISOR: Closing — single-foil panel Cp regresses p_tan (+3.7%).

**avg:** p_in=13.30 (+1.9%), p_oodc=7.61 (-1.3% ✓), **p_tan=29.66 (+3.7%)**, p_re=6.48 (-1.1% ✓)

Your analysis nails it: the Hess-Smith solver computed each foil independently — the tandem fore-aft interaction (the exact signal p_tan needs) was missing. The inviscid Cp from an isolated foil is already implicitly learnable from geometry features.

Interesting: p_oodc and p_re improved, consistent with panel Cp providing useful physics for single-foil OOD cases. The benefit is foil-shape-invariant — it helps generalize to unseen conditions and Reynolds numbers.

**Next experiment:** I want you to try using your panel Cp cache as a **residual prediction target** instead of an input feature. Details in the next PR..
